### PR TITLE
(CDAP-4253) Added authorization enforcement in Namespace HTTP APIs

### PR DESCRIPTION
**Pull request by bdmogal**
- Added a new ``EntityId`` - ``InstanceId`` to represent a CDAP instance
- Namespace create needs *ADMIN* privilege on the CDAP instance
- Namespace delete needs *ADMIN* privilege on the namespace being deleted
- Also added initial support for authorization in ``TestBase``. Currently only supports granting and revoking privileges
- Added support for listing and retrieving namespaces as well as retrieving ``CConfiguration`` in ``TestBase``
- Refactored ``CLIMainTest`` to pull out Authorization CLI Tests into a separate class which can enable authorization

Overall, the PR looks large, but mostly involves adding functionality to ``TestBase`` and refactoring some tests. The core logic of enforcement is contained in ``DefaultNamespaceAdmin``, while test is in ``AuthorizationTest``

Jira: [CDAP-4253](https://issues.cask.co/browse/CDAP-4253)
Build: http://builds.cask.co/browse/CDAP-DUT3742

### CODE REVIEW

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    CLIConfig cliConfig = createCLIConfig(STANDALONE.getBaseURI());
+    LaunchOptions launchOptions = new LaunchOptions(LaunchOptions.DEFAULT.getUri(), true, true, false);
+    CLIMain cliMain = new CLIMain(launchOptions, cliConfig);
+    cli = cliMain.getCLI();
+    testCommandOutputContains(cli, "connect " + STANDALONE.getBaseURI(), "Successfully connected");
+  }
+
+
+  @Test
+  public void testAuthorizationCLI() throws Exception {
```
Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: this was moved from ``CLIMainTest``

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLITestBase.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.net.URI;
+import javax.annotation.Nullable;
+
+/**
+ * Base class for CLI Tests.
+ */
+public class CLITestBase {
```
Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/CLITestBase.java]: these methods were moved from ``CLIMainTest``

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 3]
```diff
@@ -138,22 +155,26 @@ public boolean exists(Id.Namespace namespaceId) {
...
-    Id.Namespace namespace = Id.Namespace.from(metadata.getName());
-    if (exists(Id.Namespace.from(metadata.getName()))) {
-      throw new NamespaceAlreadyExistsException(namespace);
+    NamespaceId namespace = new NamespaceId(metadata.getName());
+    if (exists(namespace.toId())) {
+      throw new NamespaceAlreadyExistsException(namespace.toId());
     }
 
+    // Namespace can be created. Check if the user is authorized now.
+    Preconditions.checkState(authorizerInstantiatorService.isRunning(), "Authorization Service has not yet started");
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: centralize this check and error msg?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2014-2015 Cask Data, Inc.
+ * Copyright © 2014-201 Cask Data, Inc.
```
Comment (anwar6953) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java]: 2016*

File: cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java [hunk 5]
```diff
@@ -334,10 +337,35 @@ public void deleteNamespace(Id.Namespace namespace) throws Exception {
   }
 
   @Override
+  public void deleteNamespaceData(NamespaceId namespace) throws Exception {
```
Comment (anwar6953) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: why not just a `deleteNamespace` method?

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import co.cask.cdap.proto.id.EntityIdCompatible;
 import co.cask.cdap.proto.id.FlowletId;
 import co.cask.cdap.proto.id.FlowletQueueId;
+import co.cask.cdap.proto.id.InstanceId;
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: unused?

File: cdap-standalone/src/test/java/co/cask/cdap/StandaloneTester.java [hunk 7]
```diff
@@ -47,10 +45,14 @@
   private static final Logger LOG = LoggerFactory.getLogger(StandaloneTester.class);
 
   private final TemporaryFolder tmpFolder = new TemporaryFolder();
+  private final Object [] configs;
   private CConfiguration cConf;
   private StandaloneMain standaloneMain;
 
-  public StandaloneTester() {
+  public StandaloneTester(Object ... configs) {
```
Comment (anwar6953) [inline cdap-standalone/src/test/java/co/cask/cdap/StandaloneTester.java]: Why not make it `String... configs`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 8]
```diff
@@ -138,22 +155,26 @@ public boolean exists(Id.Namespace namespaceId) {
...
+    authorizerInstantiatorService.get().enforce(instanceId, getPrincipal(), Action.ADMIN);
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Will this throw an exception if the principal doesn't have admin access on the instanceId?

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Yup, it'll throw an ``UnauthorizedException``, which gets translated into an HTTP ``FORBIDDEN`` in ``HttpExceptionHandler``

File: cdap-standalone/src/test/java/co/cask/cdap/StandaloneTester.java (hunk 7 above)
Comment (bdmogal) [inline cdap-standalone/src/test/java/co/cask/cdap/StandaloneTester.java]: Followed the same convention as in ``TestConfiguration``. I think the idea is that callers don't have to convert values to strings before passing it here. They are converted inside here.

File: cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java (hunk 5 above)
Comment (bdmogal) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: coz it deletes datasets inside the namespace - not the namespace itself

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  @ClassRule
+  public static final TestConfiguration TEST_CONFIG = new TestConfiguration(
+    Constants.Security.ENABLED, true,
+    Constants.Security.Authorization.ENABLED, true,
+    Constants.Security.Authorization.EXTENSION_JAR_PATH, extensionJar.toURI().getPath()
+  );
+
+  @BeforeClass
+  public static void setup() {
+    SecurityRequestContext.setUserId("alice");
```
Comment (gokulavasan) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: constant string for 'alice'? since I guess it should be the same used in line 60?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      Assert.fail("Namespace delete datasets should have failed because alice's admin privileges have been revoked");
+    } catch (UnauthorizedException expected) {
+      // expected
+    }
+    // grant privileges again
+    getAuthorizationManager().grant(namespace, alice, ImmutableSet.of(Action.ADMIN));
+    deleteNamespaceDatasets(namespace);
+    // deleting datasets does not revoke privileges.
+    updateNamespaceProperties(new NamespaceMeta.Builder(NamespaceMeta.DEFAULT).setDescription("new desc").build());
+    deleteNamespace(namespace.toId());
```
Comment (gokulavasan) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: Can we check if the privilege has been revoked here after deletion of namespace?

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  @AfterClass
+  public static void cleanup() throws Exception {
+    // we want to execute TestBase's @AfterClass before unsetting userid
```
Comment (gokulavasan) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: Any reason why we want to execute TestBase's @AfterClass before unsetting userid?

**Comment by gokulavasan** [comment]
@bdmogal Reviewed. Minor comments.

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java (hunk 11 above)
Comment (bdmogal) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: Unsetting the userId seemed like a good thing to do, so we return to the same state we were in at the beginning of the test. It should be done after TestBase's ``@AfterClass``, so the cleanup can happen while alice is still ADMIN - in ``@BeforeClass``, we give alice ADMIN privileges to the ``default`` namespace. 

However, to make it even better, we should reset the user id to the one that existed before this test, instead of resetting it to null. Will fix.

File: cdap-proto/src/main/java/co/cask/cdap/proto/Id.java (hunk 6 above)
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/Id.java]: yeah will revert. Had added it because the ``EntityId`` classes enforce that you have an old version of the class in ``Id`` as well. I relaxed that limitation (for newer Ids), but forgot to revert this change.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java]: Right :-)

File: cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java (hunk 5 above)
Comment (anwar6953) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: It seems a bit specific to have this functionality here (also exposed to the user when they're writing their test cases).
Would they ever need to call this?

Comment (bdmogal) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: I guess only tests that test Namespace APIs. But we should make these available from somewhere. Perhaps not from ``TestBase``, but from a ``NamespaceManager``? What do you think?

Comment (bdmogal) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: I'm going to deprecate the ``createNamespace`` and ``deleteNamespace`` methods and instead expose the ``NamespaceAdmin`` via a static getter, just like ``IntegrationTestBase``.

**Comment by bdmogal** [comment]
@gokulavasan @anwar6953 addressed comments. Please review again. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 12]
```diff
@@ -271,12 +301,28 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
...
+  private InstanceId createInstanceId(CConfiguration cConf) {
+    String instanceName = cConf.get(Constants.Security.Authorization.INSTANCE_NAME);
+    if (instanceName == null) {
+      instanceName = cConf.get(Constants.Router.ADDRESS);
+    }
+    return InstanceId.fromHostName(instanceName);
+  }
+
+  private void ensureAuthorizationServiceIsRunning() {
+    Preconditions.checkState(authorizerInstantiatorService.isRunning(), "Authorization Service has not yet started");
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Can we just move this into the supplier#get method?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 13]
```diff
@@ -271,12 +301,28 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
...
   }
+
+  private Principal getPrincipal() {
+    return new Principal(SecurityRequestContext.getUserId(), Principal.PrincipalType.USER);
+  }
+
+  private InstanceId createInstanceId(CConfiguration cConf) {
+    String instanceName = cConf.get(Constants.Security.Authorization.INSTANCE_NAME);
+    if (instanceName == null) {
+      instanceName = cConf.get(Constants.Router.ADDRESS);
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: should we also include port as part of the instance name, to ensure uniqueness?
Technically, there could be two CDAP instances' routers on the same host?

**Comment by anwar6953** [comment]
just 2 comments/questions. remainder lgtm

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java (hunk 13 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: good point!

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: I'm not so sure about this approach. It needs some more thinking. I'll investigate and update this PR when I'm ready. Have addressed all other comments so far.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 14]
```diff
@@ -271,12 +297,24 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
...
       }
     }, ProgramType.values());
   }
+
+  private Principal getPrincipal() {
+    return new Principal(SecurityRequestContext.getUserId(), Principal.PrincipalType.USER);
+  }
+
+  private InstanceId createInstanceId(CConfiguration cConf) {
+    String instanceName = cConf.get(Constants.Security.Authorization.INSTANCE_NAME);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: The `INSTANCE_NAME` config shouldn't belong to `Authorization` in the config. It should be generic top level one (like namespace).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 15]
```diff
@@ -271,12 +297,24 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
...
+    if (instanceName == null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Can it be null? Maybe we should always default it to something in cdap-default.xml

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Tests authorization CLI commands. These tests are in their own class because they need authorization enabled.
+ */
+public class AuthorizationCLITest extends CLITestBase {
+  private static Location extensionJar;
+  static {
+    LocalLocationFactory locationFactory = new LocalLocationFactory(Files.createTempDir());
```
Comment (chtyim) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: no no no. Don't use Files.createTempDir in test. Why can't use a `@ClassRule` and temporary folder?

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    try {
+      extensionJar = AppJarHelper.createDeploymentJar(locationFactory, InMemoryAuthorizer.class);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
Comment (chtyim) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: throwing exception in a static initializer block will give you some very strange error (because that happen at classloading phase). Why can't it be done with `@BeforeClass`?

File: cdap-common/src/main/resources/cdap-default.xml [hunk 18]
```diff
@@ -1487,6 +1487,15 @@
   </property>
 
   <property>
+    <name>security.authorization.instance.name</name>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: as I mentioned above, it shouldn't be grouped under "security".

File: cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java [hunk 19]
```diff
@@ -135,7 +145,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(TestBase.class);
 
   @ClassRule
-  public static TemporaryFolder tmpFolder = new TemporaryFolder();
+  public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();
```
Comment (chtyim) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java]: This change is not backward compatible (public field) and TestBase is a public class.

Comment (chtyim) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java]: You can have
```java
public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();

@Deprecated
public static TemporaryFolder tmpFolder = TMP_FOLDER;
```

File: cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+import java.io.IOException;
+
+/**
+ * Unit tests with authorization enabled.
+ */
+public class AuthorizationTest extends TestBase {
+  private static Location extensionJar;
+  static {
+    LocalLocationFactory locationFactory = new LocalLocationFactory(Files.createTempDir());
```
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java]: Don't do this. See my comments above.

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java (hunk 16 above)
Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: Had to do it because I want to use the ``@ClassRule StandaloneTester``, and want to create this extension jar before that. If I use a ``TemporaryFolder`` here, there is no guarantee that it is created before the ``@ClassRule StandaloneTester``. It always fails with "The temporary folder has not yet been created". However, maybe I can make it a non-final ``@ClassRule StandaloneTester`` and instantiate it in ``@BeforeClass``. Not sure how that will work. Will try.

Comment (chtyim) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: Use `ExternalResource` and wrap the `TemporaryFolder` and `StandaloneTester` inside.

Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: I've updated it to create a ``@ClassRule`` that extends ``StandaloneTester``, instead of an ``ExternalResource`` that wraps ``TemporaryFolder`` and ``StandaloneTester``. A couple of reasons for doing this:
1. Can't call ``StandaloneTester``'s ``before()`` and ``after()`` because they are protected.
2. Don't want to replicate logic from ``StandaloneTester``, so I still have a static method that creates the extension jar and passes it to the super class' constructor.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java (hunk 19 above)
Comment (bdmogal) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java]: Thanks for catching. Fixed.

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Tests authorization CLI commands. These tests are in their own class because they need authorization enabled.
+ */
+public class AuthorizationCLITest extends CLITestBase {
+
+  private static final class StandaloneTesterWithAuthorization extends StandaloneTester {
+    private static final TemporaryFolder tempFolder = new TemporaryFolder();
```
Comment (chtyim) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: I think you'll need to call `tempFolder.create()` in the `ExternalResource.before` method.

Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: The ``tempFolder`` is not used in the parent, so I didn't override? Should we do it just as good practice?

Comment (chtyim) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: It's better be like this:

```java
private static final class AuthorizationTester extends ExternalResource {
  private final TemporaryFolder tempFolder = new TemporaryFolder();
  private StandaloneTester standaloneTester;

  protected void before() {
    tempFolder.before();
    standaloneTester = new StandaloneTester(getAuthConfigs(tempFolder.newFolder()));
    standaloneTester.before();
  }

  protected void after() {
    standaloneTester.after();
    tempFolder.after();
  }
}
```

**Comment by chtyim** [comment]
Just one comment left.

File: cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java (hunk 21 above)
Comment (bdmogal) [inline cdap-cli-tests/src/test/java/co/cask/cdap/cli/AuthorizationCLITest.java]: I tried this earlier since its much nicer, but ``before()`` and ``after()`` are protected methods so can't be accessed on the ``standaloneTester`` object.

**Comment by awholegunch** [comment]
cdap-common/src/main/resources/cdap-default.xml LGTM.

**Comment by chtyim** [comment]
LGTM

**Comment by bdmogal** [comment]
[Docs Quick Build](http://builds.cask.co/browse/CDAP-DOB156-2)
[cdap-site Appendix](http://builds.cask.co/artifact/CDAP-DOB156/shared/build-2/Docs-HTML/3.4.0-SNAPSHOT/en/admin-manual/appendices/cdap-site.html?highlight=cdap%20site%20xml)

**Comment by bdmogal** [comment]
Build passed - http://builds.cask.co/browse/CDAP-DUT3742-13. Merging.
